pkg/traceutil/trace.go

```diff
@@ -0,0 +1,60 @@
+package traceutil
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	"github.com/coreos/pkg/capnslog"
+	"go.uber.org/zap"
+)
+
+var (
+	plog = capnslog.NewPackageLogger("go.etcd.io/etcd", "trace")
+)
+
+type Trace struct {
+	operation string
+	startTime time.Time
+	steps     []step
+}
+
+type step struct {
+	time time.Time
+	msg  string
+}
+
+func New(op string) *Trace {
+	return &Trace{operation: op, startTime: time.Now()}
+}
+
+func (t *Trace) Step(msg string) {
+	t.steps = append(t.steps, step{time: time.Now(), msg: msg})
+}
+
+// Log dumps all steps in the Trace
+func (t *Trace) Log(lg *zap.Logger) {
+
+	var buf bytes.Buffer
+
+	buf.WriteString(fmt.Sprintf("The tracing of %v request:\n", t.operation))
+
+	buf.WriteString("Request started at:")
+	buf.WriteString(t.startTime.Format("2006-01-02 15:04:05"))
+	buf.WriteString(fmt.Sprintf(".%06d", t.startTime.Nanosecond()/1000))
+	buf.WriteString("\n")
+	lastStepTime := t.startTime
+	for i, step := range t.steps {
```

Comment:
k8s only trace lines for steps that are above a "step duration threshold". Worth doing here to reduce log volume?

Reply:
good idea. But we need clear documentation to avoid user confusion about missing steps in the output.

Reply:
Should we have a general step threshold or different threshold for different kinds of step?

Reply:
Maybe start with something simple and generic, similar to K8s trace.